### ISSUE TITLE
add wait_template functionality

### ISSUE THREAD

**Issue by bbrendon** [AppDaemon/appdaemon#580]
I have some yaml automations using a single wait_template and some with multiple wait_templates in a row. It seems to be possible to code a single wait_template in appD but I started stacking them and the code in appD got very complicated and hard to follow. 

Would it be possible to add a wait_template feature to appdaemon? I have a few automations that have 2-3 wait_templates in a row and migrating them to appD (with my python skills), doesn't seem to make sense.

hass wait_template docs:  https://www.home-assistant.io/docs/scripts/

**Comment by Odianosen25** [comment]
Hello @bbrendon,

Yeah understand what you want to do, and its been one of those things we been looking for ways to add. Presently due to the threaded nature of AD, having such a wait function, would block threads and will mess AD up. 

To add these waits, some have found unique ways of doing it like @ReneTode, who uses a series of `run_in` if I am not wrong. But the main feature that is being worked on, is the ability to develop asyncio based apps. Once that has been implemented, it will be possible to easily add such wait functions.

Regards

**Comment by ReneTode** [comment]
everything you can do with those scripts can be done with the existing functions from AD.

wait for a certain state -> listen_state -> do the stuff you want in a callback
wait for a certain time -> run_in -> do the stuff you want in a callback

you can use as much from those as you like in an app
for those basic scripts you probably not even need to get creative.
with things that you stacked, you will probably find out that you can do it way more easy in AD.

when you dont know how, just ask on the chat.

**Comment by bbrendon** [comment]
To the point of complexity....

The appdaemon code and automation here are similar.  The main different is that the appdaemon one has some extra checks in the "set_say_yes_no" function.
https://gist.github.com/bbrendon/584ecac7dbdabae74a8c46656e5e8e59

I'm sure there is a better way but I haven't figure one out.

**Comment by ReneTode** [comment]
i cant find anything in the docs about if the waits you placed in the automation are waits in the order that they are placed in, or that in this case both criteria need to be true.
but from your app i expect that both criteria need to be met.
so something like:
```
import appdaemon.plugins.hass.hassapi as hass
import datetime


class  notify_humidifier(hass.Hass):

    def initialize(self):
        self.counter = False
        self.motion = False
        runtime = datetime.time(22, 15, 0)
        self.run_daily(self.start_action, runtime) # your trigger

    def start_action(self,kwargs):
        if int(self.get_state("sensor.master_bed_motion_count_15min")) > 2: # wait 1
            self.counter = True
        if self.get_state("binary_sensor.master_bed_motion") == "on": # wait 2 
            self.motion = True
        if not self.counter or not self.motion:  # still waiting
            self.run_in(self.start_action,1)
        else: # wait is over do action
            self.run_in(self.notification,0)

       def notification(self,kwargs):
           # reset everything
           self.motion = False
           self.counter = False
           # now do the notify stuff
           if float(self.get_state("sensor.dark_sky_overnight_low_temperature_0d")) < 51:
               say = "Make sure to turn on the humidifier."
           else:
               say = "The humidifier won't be needed tonight."

           # If its warm tomorrow, lets just skip it and say we already said something.
           if not float(self.get_state("sensor.dark_sky_daytime_high_apparent_temperature_1d")) > 73:
               self.call_service('notify/alexa_media', target="group.alexa_boxes", message=say, data={ "type": "announce" } )
```
the problem is that even if we add a wait function it would be like:
self.wait(self.cb,entity,state)
and when the wait is over the action gets done (which would be identical to listen_state)
there is no way that its possible to create a list of waits. unless we start creating something like:
self.wait(self.cb,{entity1:state1,entity2,state2}) or self.listen_state(self.cb,{entity1:{"new":state1},entity2,{"new":state2}})
that would create very complicated code

but my code with the run_in loop can be adapted for 3, 4 or 5 criteria.
and depending on the severity you can set the delay higher.
and for motion detectors that are off after a second you could set the flags with listen_state
you could also work with 1 flag callback like:
```
    self.listen_state(self.callback, "sensor.master_bed_motion_count_15min")
    self.listen_state(self.callback, "binary_sensor.master_bed_motion", old="off", new="on")

def callback(self, entity, attribute, old, new, kwargs):
    # create if statetements if needed (only for bigger or smaller then)
    self.flag[entity] = True
```
and in the loop check if all criteria are met.

@Odianosen25  and @acockburn i hardly have use for it, but i can imagine that it would make live from people more easy if we could have something like
self.listen_state(self.cb,entity,new="5", comparison = ">")
which would check for state > 5

**Comment by Odianosen25** [comment]
@bbrendon,

I am sorry I didn't initially fully grasp what your intentions were, but based on what Rene said, you can easily achieve what you want using [custom constraints](https://appdaemon.readthedocs.io/en/latest/APPGUIDE.html#custom-constraints).

This way you can use the constraint which will grantee that the callback will not be fired, until the conditions have been met. So like using Rene's example:
```
    self.register_constraint("state_greater_than")
    entity = "sensor.master_bed_motion_count_15min"
    self.listen_state(self.callback, entity)
    self.listen_state(self.callback, "binary_sensor.master_bed_motion", old="off", new="on", state_greater_than = {entity: 5})

def callback(self, entity, attribute, old, new, kwargs):
    # create if statetements if needed (only for bigger or smaller then)
    self.flag[entity] = True

def state_greater_than(self, data):
    entity = list(data.keys())[0]
    value = list(data.values())[0]

    if float(self.get_state(entity)) > value:
         return True
    else:
         return False 
```
With the above, that callback will not be fired for `sensor.master_bed_motion_count_15min`, unless its value is greater than `5`.

There is plan to make declared constraints available across all apps, so once that is done, `state_greater_than` can then be used across all apps.

Hope it helps? Regards

**Comment by bbrendon** [comment]
Thank you guys for the responses. I'm going to spend some time this weekend digesting your ideas.

**Comment by ReneTode** [comment]
good suggestion @Odianosen25 didnt think of that way.
but i still think that if its that easy possible with a constraint, its also possible to add that.
we have all kind of constraints implemented, so why not add greater_then, smaller_then, etc.
or 1 where > and < can be given in a string.

we only need to check if its indeed a float or int. if not its gives an (user) error

**Comment by bbrendon** [comment]
I think I under what you two are doing. I think I can grasp the two examples fully if I spent time rewriting it.  It seems though wait_template in yaml is still way more elegant. It preserves order, pauses execution, has timeout overrides. Much more powerful and much easier.

Maybe my problem is that I think I should be able to pause execution and that's not how programming works.

Anyway. If I ever do this again, I'm going to revisit this thread.

**Comment by ReneTode** [comment]
> It preserves order,

yeah thats what i expected, but you code doesnt.
if you want waits in order you need different programming, and way more simple.
more like:
```
def init():
  self.run_daily(cb_1)
def cb_1(...):
  self.listen_state(cb_2, entity=entity1, new="something",oneshot = True)
def cb_2(...):
   self.listen_state(cb_3, entity=entity2, new="something",oneshot = True)
def cb_3(...):
```
that waits for entity1, untill that is done then it waits for entity2 then it runs cb_3
you can create an app like that and put all args in yaml, and you could even let it look like the wait from HA.
dont forget that what you do here in programming, is what HA does in the background from the yaml.

its ceretainly not more powerfull (because you can do exactly the same in AD)
its more easy because someone else did the programming for you. like when i write an app and you use it. but this is more flexible, because you can change it and add things you cant add to the HA yaml.

 
> Maybe my problem is that I think I should be able to pause execution and that's not how programming works.
you cant pause programming, but you can have it wait untill something happens. you just need to rethink how to build apps.
like i said: wait time == AD schedular, wait state == listen_state

you only need to jump from callback to callback

**Comment by Odianosen25** [comment]
Hello @bbrendon,

When asyncio apps has been developed, then possibly state constraint functions can be created, which can be used to do what you want.

Regards

**Comment by Odianosen25** [comment]
Apologies mistakenly closed it, instead of just commenting

**Comment by rafadesu** [comment]
I reckon that now that we have async functions it would be brilliant to have an awaitable wait_for_state or wait_for_event (with timeouts)

Maybe that is similar to what you guys are discussing here. (I can raise a different feature request if you feel that is not similar enough)

That would allow to write handling functions that could be very readable:

Save current volume of speaker
Set volume of speaker
Play message
Wait for state to change to pause
Restore volume

I currently doing it using futures but it forces me to have a callback that just sets the future result. And btw, I cannot send futures as kwargs because they are “deep copied” which is not allowed (apparently)...

I can post a more detailed example if you want.

**Comment by ReneTode** [comment]
i dont think we will start adding async functions anytime soon.
writing apps in async is recommended only for very experienced users that really know what they do, and that know enough about async.

writing apps in sync is way more save.
if you got a lot of apps doing different things and you get a problem in 1 app, in almost all cases all other apps keep on running, so a small mistake wont effect all your automations.
when you write apps in async, then every failure will effect ALL other apps. every error can stop all of AD and all apps.

if we start adding things like wait_for_state, all inexperienced users that have no idea about working with async will also start using those functions, which will cause that lots of people will run into trouble.

and for all the use cases you give there are sync options.

**Comment by rafadesu** [comment]
I guess you are right. I understand.

I can create my own versions of that easily anyways. Thanks for considering.

**Comment by bbrendon** [comment]
I guess when the next version is released this can be closed?
https://appdaemon.readthedocs.io/en/dev/APPGUIDE.html#sequence-wait-state

**Comment by Odianosen25** [comment]
Actually @bbrendon,

Will be closing this now since its in dev. Hopefully @acockburn will initiate a release soon.
